chain/chain/src/store/latest_witnesses.rs

```diff
@@ -0,0 +1,149 @@
+    /// `LatestWitnessesKey` has custom serialization to ensure that the binary representation
+    /// starts with big-endan height and shard_id.
+    /// This allows to query using a key prefix to find all witnesses for a given height (and shard_id).
+    fn serialized(&self) -> [u8; 32] {
+        let mut result = [0u8; 32];
+        result[..8].copy_from_slice(&self.height.to_be_bytes());
+        result[8..16].copy_from_slice(&self.shard_id.to_be_bytes());
+        result[16..].copy_from_slice(&self.random_uuid);
+        result
+    }
+}
+
+/// Keeps aggregate information about all witnesses stored in `DBCol::LatestChunkStateWitnesses`.
+/// Used for enforcing limits on the number of witnesses stored in the database.
+#[derive(Debug, Clone, Copy, BorshSerialize, BorshDeserialize, PartialEq, Eq, Default)]
+struct LatestWitnessesInfo {
+    pub count: u64,
+    pub total_size: u64,
+}
+
+impl LatestWitnessesInfo {
+    pub fn is_within_limits(&self) -> bool {
+        self.count <= LATEST_WITNESSES_MAX_COUNT
+            && self.total_size <= LATEST_WITNESSES_MAX_SIZE.as_u64()
+    }
+}
+
+impl ChainStore {
+    /// Saves an observed `ChunkStateWitness` to the database for later analysis and debugging.
+    /// The witness is stored in `DBCol::LatestChunkStateWitnesses`.
+    /// This function does a read-before-write. Don't call it in parallel on the same database,
+    /// or there will be race conditions.
+    pub fn save_lateset_chunk_state_witness(
+        &mut self,
+        witness: &ChunkStateWitness,
+    ) -> Result<(), std::io::Error> {
+        let serialized_witness = borsh::to_vec(witness)?;
+        let serialized_witness_size: u64 =
+            serialized_witness.len().try_into().expect("Cannot convert usize to u64");
+
+        if serialized_witness_size > SINGLE_LATEST_WITNESS_MAX_SIZE.as_u64() {
+            tracing::warn!(
+                "Cannot save latest ChunkStateWitness because it's too big. Witness size: {} byte. Size limit: {} bytes.",
+                serialized_witness.len(),
+                SINGLE_LATEST_WITNESS_MAX_SIZE.as_u64()
+            );
+        }
+
+        // Read the current `LatestWitnessesInfo`, or create a new one if there is none.
+        let mut info = self
+            .store()
+            .get_ser::<LatestWitnessesInfo>(DBCol::Misc, LATEST_WITNESSES_INFO)?
+            .unwrap_or_default();
+
+        // Adjust the info to include the new witness.
+        info.count += 1;
+        info.total_size += serialized_witness.len() as u64;
+
+        // Go over witnesses with increasing (height, shard_id) and remove them until the limits are satisfied.
+        // Height and shard id are stored in big-endian representation, so sorting the binary representation is
+        // the same as sorting the integeres.
+        let mut store_update = self.store().store_update();
+        for item in self.store().iter(DBCol::LatestChunkStateWitnesses) {
+            if info.is_within_limits() {
```

Comment:
I do not get this, does it mean that once the new witness is written it can exceed the limits and we will check is when adding another witness later?

Reply:
thanks for the explanation, I missed part of the code thinking you did not include the new witness in the computation. all looks fine.